R/aggregate.R

```diff
@@ -180,43 +143,380 @@ agg_climate_data <- function(year, data_raster, climate_var, daily_agg, trans =
+    i <- i + 1
+  }
+
+
+
+  # Aggregate by polygon
+  sum_by_poly <- after_trans(clim_dt, weights_dt, list_names, second_weights, agg_to)
+
+  return(sum_by_poly)
+}
+
+
+
+#' @export
+staggregate_bin <- function(data_raster, climate_var, daily_agg, num_bins = 30, binwidth = NULL, center_on = NULL, start_on = NULL, end_on = NULL, max = NULL, min = NULL, geoweights_table, second_weights, agg_to = "month"){
+  # Get climate data as a data table and aggregate to daily values
+  setup_list <- before_trans(data_raster, climate_var, daily_agg, geoweights_table, second_weights)
+  clim_daily <- setup_list[[1]]
+  weights_dt <- setup_list[[2]]
+  layer_names <- setup_list[[3]]
+
+
+  clim_daily_table <- raster::values(clim_daily)
+
+  if(is.null(max)){
+    max <- max(clim_daily_table) + .01
+  }
+  if(is.null(min)){
+    min <- min(clim_daily_table) - .01
+  }
+
+  # Write message that binwidth overrides num_bins
+  if(!is.null(binwidth)){
+    message(crayon::yellow("Binwidth argument supplied will override num_bins"))
+    num_bins <- ceiling((max - min) / binwidth)
+  }
+
+  # If value not supplied to binwidth, calculate from num_bins
+  if(is.null(binwidth)){
+    binwidth <- (max - min) / num_bins
+  }
+
+  # Stop with message that only one of center_on, start_on, and end_on can be chosen if otherwise
+  if((!is.null(center_on) & !is.null(start_on)) |
+     (!is.null(center_on) & !is.null(end_on)) |
+     (!is.null(start_on) & !is.null(end_on))){
+    stop(crayon::red("Too many bin placement arguments specified. Please specify a value for only one variable out of center_on, start_on, or end_on"))
+  }
+
+  if(is.null(center_on) & is.null(start_on) & is.null(end_on)){
+    message(crayon::yellow("No bin placement argument specified. Drawing bins from min value."))
+    start_on <- min
+  }
+
+  # Put start_on and end_on in terms of center_on
+  if(!is.null(start_on)){
+    center_on <- start_on + (binwidth / 2)
+  }
+
+  if(!is.null(end_on)){
+    center_on <- end_on - (binwidth / 2)
+  }
+
+
+
+  # Check to see that center_on falls within the range of data, move it if not
+  if(max - min < binwidth){
+    stop(crayon::red("The binwidth is larger than the range of data. Please specify a different binwidth and rerun"))
+  }
+
+  if(center_on > max){
+    message(crayon::yellow("Bin center is greater than maximum value in data. Adjusting it by a multiple of the binwidth so that the bin center falls within the range of data"))
+
+    while(center_on > max){
+      center_on <- center_on - binwidth
+    }
+  }
+  if(center_on < min){
+    message(crayon::yellow("Bin center is less than minimum value in data. Adjusting it by a multiple of the binwidth so that the bin center falls within the range of data"))
+
+    while(center_on < min){
+      center_on <- center_on + binwidth
+    }
+  }
+
+
+  # Create table of bins
+  center <- c(center_on)
+  while(max(center) + (binwidth / 2) < max){
+    center <- c(center, (max(center) + binwidth))
```

Comment:
shouldn't this also be `+binwidth/2`?

Reply:
Actually I think  `+ binwidth` is what we want here. This while loop and the one after it are creating a table of all the bins' centers. Adding half a binwidth would create bins centered on each others' edges. The reason we do this until the highest center is one half bin or closer toe the max value is so that the last bin we create is the one that captures max. (Though now that I think about it this needs to be a <= not a <